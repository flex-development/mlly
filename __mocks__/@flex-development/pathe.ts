/**
 * @file Mocks - @flex-development/pathe
 * @module mocks/flex-development/pathe
 * @see https://github.com/flex-development/pathe
 */

/**
 * [`@flex-development/pathe`][1] module type.
 *
 * [1]: https://github.com/flex-development/pathe
 */
type Actual = typeof import('@flex-development/pathe')

/**
 * `@flex-development/pathe` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/pathe')

export default {
  basename: vi.fn(actual.basename),
  changeExt: vi.fn(actual.changeExt),
  dirname: vi.fn(actual.dirname),
  extname: vi.fn(actual.extname),
  format: vi.fn(actual.format),
  isAbsolute: vi.fn(actual.isAbsolute),
  join: vi.fn(actual.join),
  parse: vi.fn(actual.parse),
  relative: vi.fn(actual.relative),
  removeExt: vi.fn(actual.removeExt),
  resolve: vi.fn(actual.resolve)
}
