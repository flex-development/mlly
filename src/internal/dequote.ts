/**
 * @file Internal - dequote
 * @module mlly/internal/dequote
 */

import { trim } from '@flex-development/tutils'

/**
 * Removes quotes from the beginning and end of a string.
 *
 * @internal
 *
 * @param {string} str - String to dequote
 * @return {string} String with quotes removed
 */
const dequote = (str: string): string => trim(str).replace(/^["']|["']$/g, '')

export default dequote
