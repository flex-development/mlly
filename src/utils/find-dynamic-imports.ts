/**
 * @file findDynamicImports
 * @module mlly/utils/findDynamicImports
 */

import {
  SpecifierSyntaxKind,
  StatementKind,
  StatementSyntaxKind
} from '#src/enums'
import type { DynamicImport } from '#src/interfaces'
import dequote from '#src/internal/dequote'
import getSpecifierKind from '#src/internal/get-specifier-kind'
import validateString from '#src/internal/validate-string'
import type { NodeError } from '@flex-development/errnode'
import { DYNAMIC_IMPORT_REGEX } from '@flex-development/import-regex'
import { split, trim } from '@flex-development/tutils'

/**
 * Finds all dynamic import statements in `code`. Ignores matches in comments.
 *
 * @see {@linkcode DynamicImport}
 * @see https://regex101.com/r/PTPAvU
 * @see https://developer.mozilla.org/docs/Web/JavaScript/Reference/Operators/import
 *
 * @param {string} [code=''] - Code to evaluate
 * @return {DynamicImport[]} Dynamic import statement objects
 * @throws {NodeError<TypeError>} If `code` is not a string
 */
const findDynamicImports = (code: string = ''): DynamicImport[] => {
  validateString(code, 'code')

  return [...code.matchAll(DYNAMIC_IMPORT_REGEX)].map(match => {
    const { 0: code = '', index: start = 0, groups = {} } = match
    const {
      expression = '',
      imports = '',
      options = '',
      specifier = ''
    } = groups

    /**
     * Module specifier syntax.
     *
     * @const {SpecifierSyntaxKind} specifier_syntax
     */
    const specifier_syntax: SpecifierSyntaxKind = /^["']/.test(specifier)
      ? SpecifierSyntaxKind.STATIC
      : SpecifierSyntaxKind.DYNAMIC

    return {
      code,
      end: start + code.length,
      expression,
      imports: imports === ''
        ? []
        : split(imports.replace(/^{|}$/g, ''), ',')
          .map(trim)
          .filter(e => !!e.length),
      kind: StatementKind.IMPORT,
      options,
      specifier: dequote(specifier),
      specifier_kind: specifier_syntax === SpecifierSyntaxKind.DYNAMIC
        ? null
        : getSpecifierKind(specifier),
      specifier_syntax,
      start,
      syntax: StatementSyntaxKind.DYNAMIC
    }
  })
}

export default findDynamicImports
