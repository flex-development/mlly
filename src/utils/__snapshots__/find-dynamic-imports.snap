// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:utils/findDynamicImports > should return DynamicImport object array 1`] = `
[
  {
    "code": "await import('foo-package')",
    "end": 27,
    "expression": "import('foo-package')",
    "imports": [],
    "kind": "import",
    "options": "",
    "specifier": "foo-package",
    "specifier_kind": "bare",
    "specifier_syntax": "static",
    "start": 0,
    "syntax": "dynamic",
  },
  {
    "code": "await import("/modules/my-module.js")",
    "end": 65,
    "expression": "import("/modules/my-module.js")",
    "imports": [],
    "kind": "import",
    "options": "",
    "specifier": "/modules/my-module.js",
    "specifier_kind": "absolute",
    "specifier_syntax": "static",
    "start": 28,
    "syntax": "dynamic",
  },
  {
    "code": "await import('./foo.json', { assert: { type: 'json' } })",
    "end": 122,
    "expression": "import('./foo.json', { assert: { type: 'json' } })",
    "imports": [],
    "kind": "import",
    "options": "{ assert: { type: 'json' } }",
    "specifier": "./foo.json",
    "specifier_kind": "relative",
    "specifier_syntax": "static",
    "start": 66,
    "syntax": "dynamic",
  },
  {
    "code": "const foo = await import(bar)",
    "end": 153,
    "expression": "import(bar)",
    "imports": [
      "foo",
    ],
    "kind": "import",
    "options": "",
    "specifier": "bar",
    "specifier_kind": null,
    "specifier_syntax": "dynamic",
    "start": 124,
    "syntax": "dynamic",
  },
  {
    "code": "myModule = await import("module-used-on-server")",
    "end": 257,
    "expression": "import("module-used-on-server")",
    "imports": [
      "myModule",
    ],
    "kind": "import",
    "options": "",
    "specifier": "module-used-on-server",
    "specifier_kind": "bare",
    "specifier_syntax": "static",
    "start": 209,
    "syntax": "dynamic",
  },
  {
    "code": "myModule = await import("module-used-in-browser")",
    "end": 319,
    "expression": "import("module-used-in-browser")",
    "imports": [
      "myModule",
    ],
    "kind": "import",
    "options": "",
    "specifier": "module-used-in-browser",
    "specifier_kind": "bare",
    "specifier_syntax": "static",
    "start": 270,
    "syntax": "dynamic",
  },
  {
    "code": "import(\`/modules/module-\${index}.js\`)",
    "end": 417,
    "expression": "import(\`/modules/module-\${index}.js\`)",
    "imports": [],
    "kind": "import",
    "options": "",
    "specifier": "\`/modules/module-\${index}.js\`",
    "specifier_kind": null,
    "specifier_syntax": "dynamic",
    "start": 380,
    "syntax": "dynamic",
  },
  {
    "code": "const {
  default: myDefault,
  addFive,
  addFour,
  addThree,
  addTwo,
  squareFive,
  squareFour,
  squareThree,
  squareTwo
} = await import('./lib')",
    "end": 576,
    "expression": "import('./lib')",
    "imports": [
      "default: myDefault",
      "addFive",
      "addFour",
      "addThree",
      "addTwo",
      "squareFive",
      "squareFour",
      "squareThree",
      "squareTwo",
    ],
    "kind": "import",
    "options": "",
    "specifier": "./lib",
    "specifier_kind": "relative",
    "specifier_syntax": "static",
    "start": 422,
    "syntax": "dynamic",
  },
  {
    "code": "import("/modules/my-module.js")",
    "end": 804,
    "expression": "import("/modules/my-module.js")",
    "imports": [],
    "kind": "import",
    "options": "",
    "specifier": "/modules/my-module.js",
    "specifier_kind": "absolute",
    "specifier_syntax": "static",
    "start": 773,
    "syntax": "dynamic",
  },
]
`;
